### PR TITLE
8296318: use-def assert: special case undetected loops nested in infinite loops

### PR DESCRIPTION
**Pull request by eme64**
`PhaseCFG::verify` checks that the nodes scheduled into the blocks have "use-after-def".
https://github.com/openjdk/jdk/blob/8c472e481676ed0ef475c4989477d5714880c59e/src/hotspot/share/opto/block.cpp#L1376

As long as the control flow has no loops, this should always hold.
We make sure to not check the assert if the block-head is a `LoopNode`, and the use `n` is a `Phi`, since we may have inputs from the backedge, which would be "use-before-def" in the block-scheduling order, but that is to be expected.
https://github.com/openjdk/jdk/blob/8c472e481676ed0ef475c4989477d5714880c59e/src/hotspot/share/opto/block.cpp#L1364

**Problem**
This assumes that all loops are properly detected, if a loop was not detected the block-head may only be a `RegionNode`, and the assert can fail (see regression test). Why are not all loops detected?
During `PhaseIdealLoop::build_loop_tree`, we detect all loops, but not always attach infinite loops with their sub-loops to the loop tree, and then the loop-head `Regions` are not converted to `LoopNodes` in `PhaseIdealLoop::beautify_loops`.
This behaviour is expected, see also https://github.com/openjdk/jdk/pull/11473.
Thus, the assert fires, but it should not.

**Solution**
Add special casing to assert: also accept if the `Region` is in an infinite subgraph, and we are looking at a `Phi` as use (the def could be values from the backedge).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296318](https://bugs.openjdk.org/browse/JDK-8296318): use-def assert: special case undetected loops nested in infinite loops

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [72a2f0b1](https://git.openjdk.org/jdk/pull/11642/files/72a2f0b15c35a496ac4c658a6de09a56369ba90e)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11642/head:pull/11642` \
`$ git checkout pull/11642`

Update a local copy of the PR: \
`$ git checkout pull/11642` \
`$ git pull https://git.openjdk.org/jdk pull/11642/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11642`

View PR using the GUI difftool: \
`$ git pr show -t 11642`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11642.diff">https://git.openjdk.org/jdk/pull/11642.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/cfgnode.cpp [hunk 1]
```diff
@@ -394,6 +394,45 @@ bool RegionNode::is_unreachable_from_root(const PhaseGVN* phase) const {
...
+  ResourceMark rm;
+  Unique_Node_List worklist;
+  worklist.push(this);
+  return RegionNode::are_all_in_infinite_subgraph(worklist);
+}
+
+// Are all nodes in worklist in infinite subgraph?
+// (no path to root except through false NeverBranch exit)
+// worklist is directly used for the traversal
+bool RegionNode::are_all_in_infinite_subgraph(Unique_Node_List &worklist) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/cfgnode.cpp:409]:
```suggestion
bool RegionNode::are_all_nodes_in_infinite_subgraph(Unique_Node_List& worklist) {
```

File: src/hotspot/share/opto/block.cpp [hunk 2]
```diff
@@ -1375,7 +1375,13 @@ void PhaseCFG::verify() const {
                 }
               }
             }
-            assert(is_loop || block->find_node(def) < j, "uses must follow definitions");
+            // Uses must be before definition, except if:
+            // - We are in some kind of loop we already detected
+            // - We are in infinite loop, where Region may not have been turned into LoopNode
+            assert(block->find_node(def) < j ||
+                   is_loop ||
+                   (block->head()->as_Region()->is_in_infinite_subgraph() && n->is_Phi()),
```
Comment (chhagedorn) [inline src/hotspot/share/opto/block.cpp:1383]: I suggest to swap the order of `n->is_Phi()` and `block->head()->as_Region()->is_in_infinite_subgraph()`. But this should be a rare case anyways, so it does not really matter that much.

**Review by chhagedorn — APPROVED** [review]
That looks reasonable!

File: src/hotspot/share/opto/block.cpp (hunk 2 above)
Comment (eme64) [reply src/hotspot/share/opto/block.cpp:1383]: :+1:

**Review by vnkozlov — COMMENTED** [review]
Please, merge latest JDK so that GHA builds pass.

File: src/hotspot/share/opto/cfgnode.cpp [hunk 3]
```diff
@@ -394,6 +394,45 @@ bool RegionNode::is_unreachable_from_root(const PhaseGVN* phase) const {
   return true; // The Region node is unreachable - it is dead.
 }
 
+// Is this region in an infinite subgraph?
+// (no path to root except through false NeverBranch exit)
+bool RegionNode::is_in_infinite_subgraph() {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/cfgnode.cpp:399]: It is used only in assert(). Should it be under `#ifdef ASSERT` ?

Comment (eme64) [reply src/hotspot/share/opto/cfgnode.cpp:399]: :+1:

**Comment by eme64** [comment]
Thanks @vnkozlov @chhagedorn for the reviews and suggestions!
/integrate

**Pull request closed without merging**